File: src/repository/queuedEvent.ts
import type { JetstreamEvent } from "@atcute/jetstream"
import { db } from "../database/db.ts"

interface QueuedEvent {
  id: number
  event: JetstreamEvent
}

export const getOldestQueuedEvent = async (): Promise<
  QueuedEvent | undefined
> => {
  const result = await db
    .selectFrom("queued_events")
    .selectAll()
    .orderBy("id", "asc")
    .executeTakeFirst()

  return result
    ? {
      id: result.id,
      event: result.event_json,
    }
    : undefined
}

export const addQueuedEvent = async (event: JetstreamEvent) => {
  await db
    .insertInto("queued_events")
    .values({
      event_json: JSON.stringify(event),
    })
    .execute()
}

export const deleteQueuedEvent = async (id: number) => {
  await db
    .deleteFrom("queued_events")
    .where("id", "=", id)
    .execute()
}

File: src/repository/queued_event.rs
use serde_json::Value;
use sqlx::FromRow;

use crate::database::DbPool;

#[derive(FromRow)]
pub struct QueuedEvent {
    pub id: i64,
    pub event_json: Value,
}

pub async fn get_oldest_queued_event(pool: &DbPool) -> anyhow::Result<Option<QueuedEvent>> {
    let row = sqlx::query_as!(
        QueuedEvent,
        r#"SELECT id, event_json AS "event_json: serde_json::Value" FROM queued_events ORDER BY id ASC LIMIT 1"#
    )
    .fetch_optional(pool)
    .await?;

    Ok(row)
}

pub async fn add_queued_event(pool: &DbPool, event: &Value) -> anyhow::Result<()> {
    sqlx::query!("INSERT INTO queued_events (event_json) VALUES (?)", event)
        .execute(pool)
        .await?;

    Ok(())
}

pub async fn delete_queued_event(pool: &DbPool, id: i64) -> anyhow::Result<()> {
    sqlx::query!("DELETE FROM queued_events WHERE id = ?", id)
        .execute(pool)
        .await?;

    Ok(())
}
